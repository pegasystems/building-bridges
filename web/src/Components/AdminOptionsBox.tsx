import React, { FormEvent } from 'react';
import { decamelizeKeys } from 'humps';
import FormGroup from '@material-ui/core/FormGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox from '@material-ui/core/Checkbox';
import FormControl from '@material-ui/core/FormControl';

import { SURVEYS_API } from '../Consts';
import { TextField } from '@material-ui/core';

interface AdminOptionsProps {
    surveyKey: string;
    adminSecret: string;
    askingQuestionsEnabled: boolean;
    votingEnabled: boolean;
    limitQuestionCharactersEnabled: boolean;
    limitQuestionCharacters: number;
}

export default class AdminOptionsBox extends React.Component<AdminOptionsProps, any> {

    constructor(props: AdminOptionsProps) {
        super(props)
        this.state = {
            askingQuestionsEnabled: this.props.askingQuestionsEnabled,
            votingEnabled: this.props.votingEnabled,
            limitQuestionCharactersEnabled: this.props.limitQuestionCharactersEnabled,
            limitQuestionCharacters: this.props.limitQuestionCharacters
        }
        this.handleChange = this.handleChange.bind(this);
    }

    componentDidUpdate(prevProps: any) {
        if (prevProps !== this.props) {
            this.setState({
                askingQuestionsEnabled: this.props.askingQuestionsEnabled,
                votingEnabled: this.props.votingEnabled,
                limitQuestionCharactersEnabled: this.props.limitQuestionCharactersEnabled,
                limitQuestionCharacters: this.props.limitQuestionCharacters
            })
        }
    }

    handleChange = async (e: FormEvent<HTMLElement>) => {
        const setting: {[index: string]: any} = {}
        const target = e.target as HTMLInputElement;
        var value;
        switch(target.type) {
            case 'checkbox':
                value = target.checked;
                break;
            case 'number':
                value =  Number.parseFloat(target.value);
                break;
            default:
                value = target.value;

        }
        const name = target.name;
        this.setState({ [name]: value })
        setting[name] = value
        fetch(`${SURVEYS_API}${this.props.surveyKey}?admin_secret=${this.props.adminSecret}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(decamelizeKeys(setting))
        });
      }

    render(): JSX.Element {
        return (<div className="admin-box">
                <FormControl component="fieldset">
                    <FormGroup>
                        <FormControlLabel
                            control={
                            <Checkbox
                                checked={this.state.askingQuestionsEnabled}
                                onChange={this.handleChange} 
                                name="askingQuestionsEnabled"
                                color="primary" />
                            }
                            label="Asking Questions Enabled"/>
                        <FormControlLabel
                            control={
                            <Checkbox
                                checked={this.state.votingEnabled}
                                onChange={this.handleChange} 
                                name="votingEnabled"
                                color="primary" />
                            }
                            label="Voting Enabled"/>
                        <FormControlLabel
                            control={
                            <Checkbox
                                checked={this.state.limitQuestionCharactersEnabled}
                                onChange={this.handleChange} 
                                name="limitQuestionCharactersEnabled"
                                color="primary" />
                            }
                            label="Limit question length"/>
                        {
                            this.state.limitQuestionCharactersEnabled && <FormControlLabel
                                control={
                                <TextField
                                    value={this.state.limitQuestionCharacters}
                                    onChange={this.handleChange} 
                                    type="number"
                                    name="limitQuestionCharacters"
                                    color="primary"
                                    style={{
                                        width:((this.state.limitQuestionCharacters.toString().length + 1) * 12) + 'px'
                                    }} />
                                }
                                label="characters"/>
                        }
                    </FormGroup>
                </FormControl>
            </div>
        );
    }
}
