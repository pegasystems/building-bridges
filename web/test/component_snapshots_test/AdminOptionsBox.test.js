import AdminOptionsBox from '../../src/Components/AdminOptionsBox';
import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({ok: 'true'}),
  })
);

test('AdminOptions renders properly test', () => {
    const component = renderer.create(
        <AdminOptionsBox 
            open={true} 
            surveyKey={'test_survey_key'} 
            adminSecret={'admin_secret'}
        />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('AdminOptions properly handle submit', () => {
    function callback(state) {
        try {
          expect(state).toBe(false);
          done();
        } catch (error) {
          done(error);
        }
    }
    const component = shallow(
        <AdminOptionsBox 
            open={true} 
            surveyKey={'test_survey_key'} 
            adminSecret={'admin_secret'}
            surveyStateCallback={callback}
        />
    );
    component.instance().handleSubmit('')
}); 
