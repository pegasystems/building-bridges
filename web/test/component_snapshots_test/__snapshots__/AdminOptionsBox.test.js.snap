// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminOptions renders properly test 1`] = `
<div
  className="admin-box"
>
  <fieldset
    className="MuiFormControl-root"
  >
    <div
      className="MuiFormGroup-root"
    >
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          aria-disabled={false}
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorPrimary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <input
              checked={true}
              className="PrivateSwitchBase-input-4"
              data-indeterminate={false}
              disabled={false}
              name="askingQuestionsEnabled"
              onChange={[Function]}
              type="checkbox"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Asking Questions Enabled
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          aria-disabled={false}
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorPrimary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <input
              checked={true}
              className="PrivateSwitchBase-input-4"
              data-indeterminate={false}
              disabled={false}
              name="votingEnabled"
              onChange={[Function]}
              type="checkbox"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Voting Enabled
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          aria-disabled={false}
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <input
              className="PrivateSwitchBase-input-4"
              data-indeterminate={false}
              disabled={false}
              name="limitQuestionCharactersEnabled"
              onChange={[Function]}
              type="checkbox"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Limit question length
        </span>
      </label>
    </div>
  </fieldset>
</div>
`;
